geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/auth/AuthWhileServersRestartDUnitTest.java

```diff
@@ -0,0 +1,127 @@
+
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
+import org.junit.After;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import redis.clients.jedis.HostAndPort;
+import redis.clients.jedis.JedisCluster;
+
+import org.apache.geode.examples.SimpleSecurityManager;
+import org.apache.geode.internal.AvailablePortHelper;
+import org.apache.geode.redis.ConcurrentLoopingThreads;
+import org.apache.geode.test.dunit.rules.MemberVM;
+import org.apache.geode.test.dunit.rules.RedisClusterStartupRule;
+import org.apache.geode.test.dunit.rules.SerializableFunction;
+import org.apache.geode.test.junit.rules.ExecutorServiceRule;
+import org.apache.geode.test.junit.rules.ServerStarterRule;
+
+public class AuthWhileServersRestartDUnitTest {
+
+  @ClassRule
+  public static RedisClusterStartupRule clusterStartUp = new RedisClusterStartupRule(4);
+
+  @ClassRule
+  public static ExecutorServiceRule executor = new ExecutorServiceRule();
+
+  private static int redisServerPort;
+  private static SerializableFunction<ServerStarterRule> operatorForVM3;
+  private static final String KEY = "key";
+  private static final int SO_TIMEOUT = 10_000;
+
+  @BeforeClass
+  public static void classSetup() {
+    MemberVM locator = clusterStartUp.startLocatorVM(0,
+        x -> x.withSecurityManager(SimpleSecurityManager.class));
+    int locatorPort = locator.getPort();
+
+    SerializableFunction<ServerStarterRule> serverOperator = s -> s
+        .withCredential("cluster", "cluster")
+        .withConnectionToLocator(locatorPort);
+
+    clusterStartUp.startRedisVM(1, serverOperator);
+    clusterStartUp.startRedisVM(2, serverOperator);
+
+    int server3Port = AvailablePortHelper.getRandomAvailableTCPPort();
+    String finalRedisPort = Integer.toString(server3Port);
+
+    operatorForVM3 = serverOperator.compose(o -> o
+        .withProperty(GEODE_FOR_REDIS_PORT, finalRedisPort));
+
+    clusterStartUp.startRedisVM(3, operatorForVM3);
+
+    redisServerPort = clusterStartUp.getRedisPort(1);
+    // Make sure buckets get created
+    new JedisCluster(new HostAndPort(BIND_ADDRESS, redisServerPort),
+        REDIS_CLIENT_TIMEOUT, SO_TIMEOUT, DEFAULT_MAX_ATTEMPTS, "data", "data", "bootstrap",
+        new GenericObjectPoolConfig<>());
+  }
+
+  @After
+  public void after() {
+    clusterStartUp.flushAll("data", "data");
+  }
+
+  @Test
+  public void testReconnectionWithAuthAndServerRestarts() throws Exception {
+    AtomicBoolean running = new AtomicBoolean(true);
+
+    Future<?> future = executor.submit(() -> {
+      try {
+        for (int i = 0; i < 20 && running.get(); i++) {
+          clusterStartUp.moveBucketForKey(KEY, "server-3");
+          // Sleep for a bit so that commands can execute
+          Thread.sleep(2000);
```

Comment:
Is there a better way to do this than sleep? Is there something to wait for? Sleeping by itself is almost always a bad idea.

Reply:
Also, perhaps I don't understand fully, but why use a completable future here? It seems unnecessary and overly complex. It could be easily done with AsyncInvocations, which are more common in the code base.

Reply:
Good call - I've switch over to using an awaitility statement.

Regarding the future used here - `AsyncInvocation` is used for remote async calls. This code is all just local to this JVM and does not do any obvious remote calls.